File: app/src/main/java/com/aayush/filmipedia/util/api/RestApi.java
package com.aayush.filmipedia.util.api;

import com.aayush.filmipedia.model.CreditsResponse;
import com.aayush.filmipedia.model.GenreResponse;
import com.aayush.filmipedia.model.Movie;
import com.aayush.filmipedia.model.MovieCreditsResponse;
import com.aayush.filmipedia.model.MovieResponse;
import com.aayush.filmipedia.model.Person;
import com.aayush.filmipedia.model.PersonResponse;
import com.aayush.filmipedia.model.TrailerResponse;

import io.reactivex.Observable;
import retrofit2.http.GET;
import retrofit2.http.Path;
import retrofit2.http.Query;

public interface RestApi {
    @GET("movie/popular")
    Observable<MovieResponse> getPopularMovies(@Query("api_key") String apiKey,
                                               @Query("page") Long page,
                                               @Query("region") String region);

    @GET("search/movie")
    Observable<MovieResponse> searchMovies(@Query("api_key") String apiKey,
                                           @Query("query") String query,
                                           @Query("page") Long page,
                                           @Query("region") String region);

    @GET("search/person")
    Observable<PersonResponse> searchPeople(@Query("api_key") String apiKey,
                                            @Query("query") String query,
                                            @Query("page") Long page,
                                            @Query("region") String region);

    @GET("genre/movie/list")
    Observable<GenreResponse> getMovieGenres(@Query("api_key") String apiKey);

    @GET("discover/movie")
    Observable<MovieResponse> getMoviesByGenre(@Query("api_key") String apiKey,
                                               @Query("page") Long page,
                                               @Query("with_genres") int genreId,
                                               @Query("region") String region);

    @GET("person/popular")
    Observable<PersonResponse> getPopularPeople(@Query("api_key") String apiKey,
                                                @Query("page") Long page);

    @GET("movie/{movie_id}")
    Observable<Movie> getMovieById(@Path("movie_id") Long id,
                                   @Query("api_key") String apiKey);

    @GET("movie/{movie_id}/credits")
    Observable<MovieCreditsResponse> getMovieCredits(@Path("movie_id") Long id,
                                                     @Query("api_key") String apiKey);

    @GET("movie/{movie_id}/videos")
    Observable<TrailerResponse> getMovieTrailers(@Path("movie_id") Long id,
                                           @Query("api_key") String apiKey);

    @GET("person/{person_id}")
    Observable<Person> getPersonById(@Path("person_id") Long id,
                               @Query("api_key") String apiKey);

    @GET("person/{person_id}/movie_credits")
    Observable<CreditsResponse> getCreditsById(@Path("person_id") Long id,
                                               @Query("api_key") String apiKey);
}

File: app/src/main/java/com/aayush/filmipedia/util/api/RestApi.kt
package com.aayush.filmipedia.util.api

import com.aayush.filmipedia.model.*
import io.reactivex.Observable
import retrofit2.http.GET
import retrofit2.http.Path
import retrofit2.http.Query

interface RestApi {
    @GET("movie/popular")
    fun getPopularMovies(
        @Query("api_key") apiKey: String,
        @Query("page") page: Long?,
        @Query("region") region: String
    ): Observable<MovieResponse>

    @GET("search/movie")
    fun searchMovies(
        @Query("api_key") apiKey: String,
        @Query("query") query: String,
        @Query("page") page: Long?,
        @Query("region") region: String
    ): Observable<MovieResponse>

    @GET("search/person")
    fun searchPeople(
        @Query("api_key") apiKey: String,
        @Query("query") query: String,
        @Query("page") page: Long?,
        @Query("region") region: String
    ): Observable<PersonResponse>

    @GET("genre/movie/list")
    fun getMovieGenres(@Query("api_key") apiKey: String): Observable<GenreResponse>

    @GET("discover/movie")
    fun getMoviesByGenre(
        @Query("api_key") apiKey: String,
        @Query("page") page: Long?,
        @Query("with_genres") genreId: Int,
        @Query("region") region: String
    ): Observable<MovieResponse>

    @GET("person/popular")
    fun getPopularPeople(
        @Query("api_key") apiKey: String,
        @Query("page") page: Long?
    ): Observable<PersonResponse>

    @GET("movie/{movie_id}")
    fun getMovieById(
        @Path("movie_id") id: Long?,
        @Query("api_key") apiKey: String
    ): Observable<Movie>

    @GET("movie/{movie_id}/credits")
    fun getMovieCredits(
        @Path("movie_id") id: Long?,
        @Query("api_key") apiKey: String
    ): Observable<MovieCreditsResponse>

    @GET("movie/{movie_id}/videos")
    fun getMovieTrailers(
        @Path("movie_id") id: Long?,
        @Query("api_key") apiKey: String
    ): Observable<TrailerResponse>

    @GET("person/{person_id}")
    fun getPersonById(
        @Path("person_id") id: Long?,
        @Query("api_key") apiKey: String
    ): Observable<Person>

    @GET("person/{person_id}/movie_credits")
    fun getCreditsById(
        @Path("person_id") id: Long?,
        @Query("api_key") apiKey: String
    ): Observable<CreditsResponse>
}